### ISSUE TITLE
Threshold compression hangs.

### ISSUE THREAD

**Issue by acse-jx120** [sands-lab/grace#16]
When I use grace's dist package to do threshold compression by Allgather, it will freeze on communication. I found two issues on different backends.

### NCCL
Grace's Allgather will transfer tensor to [the default cuda device](https://github.com/sands-lab/grace/blob/master/grace_dl/dist/communicator/allgather.py#:~:text=local_sizes%20%3D%20torch.tensor(%5Bt.numel()%20for%20t%20in%20tensors%5D).cuda()). As I tested, NCCL cannot handle communication from the same GPU, but GLOO can do. It can run after I changed it to:  
`local_sizes = torch.tensor([t.numel() for t in tensors]).cuda(tensors[0].device.index)`

### GLOO
In an extreme situation, the threshold is bigger than all gradients, then the value and index tensors will be empty. NCCL can gather an empty list but GLOO will be stuck [here](https://github.com/sands-lab/grace/blob/master/grace_dl/dist/communicator/allgather.py#:~:text=dist.all_gather(gathered%2C%20tensor))

**Comment by hangxu0304** [comment]
Hi,

Thanks for the feedback.

The Allgather issue can be fixed in 2 ways, either explicitly setting the default device for each process, i.e. `torch.cuda.set_device(args.local_rank)` or the way you proposed here.

The Threshold issue can be fixed by forcing the threshold value to be smaller than the maximum absolute element in the input tensor, so that the value and index tensors won't be empty.

Please let us know if you have any problems.
